[FrontEnd][layer5.io] fix: Flickering of dark mode #3060

randychilau opened the pull request:
**Description**

This PR fixes #3060 

Site flickers when refreshed in dark mode:
![flicker](https://user-images.githubusercontent.com/90356410/225827119-0c7967ed-2c7b-4271-9c38-aa0a6d50b8e3.gif)


fix implemented:
![flicker_fix3](https://user-images.githubusercontent.com/90356410/225830509-e395814a-2cb6-4786-859b-07f8657c1bf1.gif)


**Notes for Reviewers**
The code in this PR borrowed heavily from the following sources:
https://github.com/gperl27/gatsby-styled-components-dark-mode 
https://www.joshwcomeau.com/react/dark-mode/
(attempted to use the above plugin and useDarkMode plugin, but experienced installation and implementation issues)

The fix involves the styled-component's ThemeProvider, Navigation (for toggling modes), and GlobalStyle. The values and functions in these components are made available to pages through the Context API, wrapRootElement, and the Gatsby layout component when required.

important files/folder:
src/app/theme
src/onRenderBody.js
src/context-wrapper.js

This method allows the removal of these components, import statements and any related code from pages in the src/pages folder. For components which have conditionally displayed elements based on dark mode (using the theme variable that was passed down via props), the useStyledDarkMode hook provides an isDark boolean to create the theme variable when necessary.

**[Signed commits](https://github.com/layer5io/layer5/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x ] Yes, I signed my commits.
 
<!--
Thank you for contributing to Layer5 projects! 

Contributing Conventions:

1. Include descriptive PR titles with [<component-name>] prepended.
2. Build and test your changes before submitting a PR. 
3. Sign your commits

By following the community's contribution conventions upfront, the review process will 
be accelerated and your PR merged more quickly.
-->

l5io commented:
🚀 Preview for commit 70999ec170cc7fe80fce7b0b61a2093a48f82a07 at: https://641429d512f90c16f6a629b2--layer5.netlify.app

Nikhil-Ladha reviewed (commented):
There are still some issue like, if you switch to light theme from dark theme and reload a couple of components like the `Card` loads intially with the dark background color. The same could be seen on the home page in the projects section as well.
Also, switch to light theme and duplicate a tab like the blog page, the same effect could be seen on first load.

randychilau commented:
Hi Nikhil, 

Thanks for the comments and taking the time to review, much appreciated.

Regarding: 
> I don't think we need this file, seems like it got duplicated by mistake.

Yes, the file is unnecessary in this case. I was debugging something regarding that file and mistakenly included it.

Regarding: 
> "if you switch to light theme from dark theme and reload a couple of components like the Card loads initially with the dark background color."

example: 
![conditional_flicker](https://user-images.githubusercontent.com/90356410/225965772-35ce6f25-dfe2-49bd-928b-0960731e12c7.gif)

I believe this is due to using props.theme.DarkTheme for conditionals in the styled-component. 

example in src/sections/Projects/Project-grid/projectGrid.style.js:
![image](https://user-images.githubusercontent.com/90356410/225964716-911612df-179d-49a1-8da4-2596bab16378.png)

The initial HTML generated on the server side will always display the false option as DarkTheme is undefined at the time. (note: this extends to any similarly related conditional for rendering e.g. the logo in Navigation).  

As stated [in this blog post by josh comeau](https://www.joshwcomeau.com/react/dark-mode/#adding-a-toggle-13) 
> "the initial render happens in the cloud at compile-time, so colorMode will initially be undefined. Every user gets the same HTML, and that HTML will always come with an unchecked checkbox."

If these type of conditionals in the styled components are changed to a props.theme value (e.g. props.theme.navigationColor), it displays correctly.

![image](https://user-images.githubusercontent.com/90356410/225961565-9b3377e0-fc2b-4f73-a564-2a3e5eb09541.png)

![conditional_flicker_fix](https://user-images.githubusercontent.com/90356410/225965862-89666f40-80bd-41be-b007-6d7244330091.gif)

note: similar conditionals for things like images have two options:
1. not rendered until color mode is known
2. display the default image and change if necessary

For the styled-component conditionals, I see 249 results in 93 files in the src/pages folder. Each styled-component has different conditionals (e.g. some for shadows, some for background color) and need to be matched to an appropriate value in the themes in src/theme/app/themeStyles.js.

Please let me know if I should update the branch with edited style-component files or create another PR.

Thanks!

randychilau commented on src/sections/Home/Banner/Banner-4/index.js:
Yes, thanks for catching this. 

I am still learning github and all the processes, is there anything I should do in this case?

Cheers.

Nikhil-Ladha replied:
Just remove it and commit the change.

randychilau replied:
Removed in branch updates, thanks!

Nikhil-Ladha commented:
Yeah, that seems like the way to go here.
It would be good to get it fixed in this PR itself. I have reviewed the changes till now so adding them to the current PR won't be an issue.

Another thing would be the toggle icon, even for intial load of dark mode the icon being displayed is "light" , probably that is again because of the conditional use. So, I will let you take care of it.

randychilau commented:
Thanks for the reply.

Sounds good, I will work on those files and update the branch.

I'll also update the toggle icon, please note there will be delay in the icon appearing on the screen.

from that [helpful post](https://www.joshwcomeau.com/react/dark-mode/#adding-a-toggle-13):
> Our best bet is to defer rendering of the toggle until after the React app knows what the colorMode should be:
> 
> By not rendering anything on the first compile-time pass, we leave a blank spot that can be filled in on the client, when that data is known to React

Nikhil-Ladha commented:
> Thanks for the reply.
> 
> Sounds good, I will work on those files and update the branch.
> 
> I'll also update the toggle icon, please note there will be delay in the icon appearing on the screen.
> 
> from that [helpful post](https://www.joshwcomeau.com/react/dark-mode/#adding-a-toggle-13):
> 
> > Our best bet is to defer rendering of the toggle until after the React app knows what the colorMode should be:
> > By not rendering anything on the first compile-time pass, we leave a blank spot that can be filled in on the client, when that data is known to React

Seems fair.
However, let's do a bit better job here.

We would render the toggle but just hide it out using CSS initially, this would help use avoid CLS (Cummulative Layout Shift) issues in lighthouse. And, then as the colormode becomes available we would render it with appropriate value.

This is not that big of a change to consideer in this PR, I am fine with you sending out a new PR for this.

Moreover, _probably_ if everything goes smoothly we would shift to the dark theme being the default theme. So, you could make both the changes in the next PR.

I am really excited for this PR, great work @randychilau 🥇 🚀

Chadha93 commented:
@randychilau  Let's discuss this on the Websites call. Please add this as an agenda item in the [meeting minutes](https://docs.google.com/document/d/1XczAHXVe2FIWPqiF57ospJ43zw5cZQ7ui8mn39v5EvA/edit#) if you would. :)

randychilau commented:
@Nikhil-Ladha 

Good day, I have updated the branch to:

- removed duplicate folder and file 

- resolve DarkTheme conditionals in [names].style.js files to prevent flash of showing wrong color when not using default theme.

- updated theme toggle to only display after color mode is known using visibility property.

note: 

- have not updated the "CONTRIBUTING.md" file which still includes using DarkTheme conditionals for contributions. Example:

> To change text color according to the theme (recommended):
> 
> ```
> color: ${props => props.theme.DarkTheme ? props.theme.secondaryColor : props.theme.primaryLightColor}; 

If these changes are accepted, then these will need to be updated.

- kept all the color pairs used for switching themes, however if new colors are used then the themeStyles.js file will need to be updated with the new color pairs. 

---------

@Chadha93 Apologies for not seeing your message in time. Please let me know how you would like to proceed.


Cheers,
Randy

l5io commented:
🚀 Preview for commit 6344edd1b07be8e6dfb9c10ffa73f1b2bf69d592 at: https://6418f472181fbb1336ddd022--layer5.netlify.app

randychilau commented:
@leecalcote Thanks, I appreciate the opportunity to contribute and how everyone here has been friendly and supportive.  Cheers!

Nikhil-Ladha reviewed (commented):
I think we both would agree that the current naming of the theme key's is not the _best_ in the world 😅 . But, to get the best results this is something that we have to live with for now.

I think we can improve it to a certain extent, but that could be a different PR for sure along with the Contributing.md and the Theme.md updates.

Rest looks good, only a small change is needed as mentioned below.

l5io commented:
🚀 Preview for commit 5029a74b9b2d228ad61fdc05033ed9e6ea982ed0 at: https://641b3ad3c508ef009f038c89--layer5.netlify.app

randychilau commented on onRenderBody.js:
```diff
@@ -0,0 +1,48 @@
+              default:
+                  colorMode = '${ThemeSetting.LIGHT}'
+            }
+          const root = document.documentElement;
+          const iterate = (obj) => {
+              Object.keys(obj).forEach(key => {
+                 if (typeof obj[key] === 'object') {
+                      iterate(obj[key])
+                 } else {
+                   root.style.setProperty("--" + key, obj[key])
+                 }
+                 })
+          }
+          const parsedTheme = JSON.parse('${JSON.stringify(props.theme)}')
+          const theme = parsedTheme[colorMode]
+          iterate(theme)
+          root.style.setProperty('--initial-color-mode', colorMode);  
```
Hi Nikhil,

Thanks for the comment and reviewing.

Please let me know if I have misunderstood your comment/request.

--------
I believe the onRenderBody code highlighted is necessary:
- to add CSS variable values to the root styles 
- so they can be referenced by a temporary CSS variable version of the default theme
- that is created in the StyledThemeProvider for the initial HTML created by Gatsby

My understanding is that this temporary CSS variable version of the theme is used while waiting for React to re-render, and is then replaced by the correct version of the theme.

If the "iterate" function is removed from onRenderBody, the flash returns, as the initialHTML will use the lighttheme by default:

![without_iterate](https://user-images.githubusercontent.com/90356410/227028732-e17a74dd-e1de-4dab-bac6-36fc62b95219.gif)

with the "iterate" function:

![with_iterate](https://user-images.githubusercontent.com/90356410/227034320-c52c1ed0-7371-40ea-8c20-ab5bbd8129cf.gif)

--------

in more detail:

The StyledThemeProvider component temporarily uses CSS variables to show the correct color mode in the initial HTML.
```
const theme = {
    ...(didLoad ? currentTheme : transformTheme(currentTheme)),
  };
```
Since didLoad is “false” in the initial HTML, "theme" will be the result of the "transformTheme" function. This function converts the values in the object into CSS-variables.
```
  Object.keys(theme).forEach((key) => {
    const value = theme[key];
    if (typeof value === "object" && !!value) {
      newTheme[key] = transformTheme(value);
    } else {
      newTheme[key] = `var(--${key})`;
    }
  });
```
example:
`background-color:  ${props => props.theme.body};`

will show in the initial HTML as:
`background-color:  var(--body)`

The background color will be whatever "--body" is set in the root style property. We set these root style properties in the onRenderBody function with the "iterate" function before the initial HTML is presented.

--------

Apologies if this explanation is not clear or I misunderstood your comment. Happy to discuss further.

Thanks,
Randy

Nikhil-Ladha replied:
Hmm. interesting. I just noticed this behaviour.
Thanks for the explanation it helped :)

l5io commented:
🚀 Preview for commit b544a09a9cda5239c82cccc97e36ffcaaf815647 at: https://641ba0300f3aae37daca6a8a--layer5.netlify.app

Nikhil-Ladha commented:
There seems to be some transparency issue on the home page's project cards, PTAL.

randychilau commented:
Hi Nikhil,

Thanks for finding the issue.

I used the incorrect syntax for the rgb codes by forgetting the slash for the alpha channel.
> Optionally it may also be given a slash / followed by a fourth value, representing alpha ([src](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value/rgb))

I updated the themeStyles.js file and the hover, box-shadow on the buttons and cards seem to be working correctly now.
![rgb_fix](https://user-images.githubusercontent.com/90356410/227287228-9a7239fa-d1f8-4b29-bfb7-3346e8028e33.gif)

Cheers,
Randy

randychilau commented:
Hi Nikhil,

Regarding the current naming of the theme keys, I agree 😅.

I couldn't think of a better naming convention for all the color codes used, in addition to various opacity/alpha settings.

Thanks,
Randy

l5io commented:
🚀 Preview for commit 97faceaa6e7ffd92797d1fe621a06fed6cc827eb at: https://641ca3bfc160b400bdc5dea4--layer5.netlify.app

Nikhil-Ladha commented:
![Screenshot 2023-03-24 at 12 09 14 PM](https://user-images.githubusercontent.com/29438875/227444599-6c779db5-7cba-4f78-bc67-e2fd47738e67.png)

This is still not fixed

l5io commented:
🚀 Preview for commit 40266ec2ea56aa4a72aefdf65d051a87151f79e7 at: https://641d5c256d814460be27310f--layer5.netlify.app

randychilau commented:
Hi Nikhil,

Another great find, I was missing the other end of a color pair that is used for that background in themeStyles.js. 

I did a check and found one other missing pair, and added both to themeStyles.js. It is displaying correctly now:

![image](https://user-images.githubusercontent.com/90356410/227456344-35d60f2f-e24e-4b40-b744-644c7abb8a5a.png)

In the text below those cards, it seemed a bit dark (it had one color for both light and dark mode, #737373). I added a color pair transition from light grey for dark mode and dark grey for light mode.

Regarding the blockquote, I changed the svg color to a light grey which seems work well for both light and dark mode at a .3 opacity:

![image](https://user-images.githubusercontent.com/90356410/227465825-c0cabaab-0ce3-4434-940b-f26db1cd2b3b.png)


Let me know if these settings should be changed.

Thanks for working with me on this.

Cheers,
Randy

l5io commented:
🚀 Preview for commit 4415df3f661b402f5283999cc6a30b89edae84ce at: https://641d62954293c666ba3f6cfc--layer5.netlify.app

l5io commented:
🚀 Preview for commit e32368eca38c5101ae6ad4e0062d3ed142bd9dfe at: https://641dee84b8a9fe25c5f39782--layer5.netlify.app

randychilau commented:
HI Nikhil,

I resolved the merge conflicts and resolved the double navigation issue.

I am new to git, and the process to fix the DCO issue using the instructions provided seems a bit confusing:

When I type in: "git rebase HEAD~42 --signoff" as stated in the[ link](https://github.com/layer5io/layer5/pull/3972/checks?check_run_id=12246420041), I see the following

![image](https://user-images.githubusercontent.com/90356410/227623685-123fa421-f9f3-4284-8ef2-da3f57ec9257.png)

However resolving this conflict, leads into more conflicts in the next commit and so on. This process did not seem correct to me, as I am not sure which changes to accept in each case, etc.

If you have any advice on how to resolve or proceed, please let me know.

Thanks,
Randy

Nikhil-Ladha commented:
> HI Nikhil,
> 
> I resolved the merge conflicts and resolved the double navigation issue.
> 
> I am new to git, and the process to fix the DCO issue using the instructions provided seems a bit confusing:
> 
> When I type in: "git rebase HEAD~42 --signoff" as stated in the[ link](https://github.com/layer5io/layer5/pull/3972/checks?check_run_id=12246420041), I see the following
> 
> ![image](https://user-images.githubusercontent.com/90356410/227623685-123fa421-f9f3-4284-8ef2-da3f57ec9257.png)
> 
> However resolving this conflict, leads into more conflicts in the next commit and so on. This process did not seem correct to me, as I am not sure which changes to accept in each case, etc.
> 
> If you have any advice on how to resolve or proceed, please let me know.
> 
> Thanks, Randy

Hey, considering the size of the PR we would have considered one unsigned commit.  I understand that git and DCO can some times be a PITA.

I see that you anyway went ahead and added those commits, but I think a lit of unrelated changes also came along with it that are already merged in the master?

Nikhil-Ladha commented:
Ok, I think it all went safely.
Though, the DCO check is still failing but never mind, we would still take it in 👍🏼

l5io commented:
🚀 Preview for commit b2a9573f1d032db993be129303e576e7cb8d2e8e at: https://641ea70de2f56d02daf7a366--layer5.netlify.app

randychilau commented:
HI Nikhil,

Thanks for your reply.

It seems to be a bit of mess with the rebase now. I would like to roll it back, but I don't think that is possible now.

I am not sure if this is helpful at all, but I have a folder with a clone of the issue3060 branch before I did the rebase, but not sure what to do with it.

Let me know how best to proceed.

Thanks,
Randy

Nikhil-Ladha commented:
> HI Nikhil,
> 
> Thanks for your reply.
> 
> It seems to be a bit of mess with the rebase now. I would like to roll it back, but I don't think that is possible now.
> 
> I am not sure if this is helpful at all, but I have a folder with a clone of the issue3060 branch before I did the rebase, but not sure what to do with it.
> 
> Let me know how best to proceed.
> 
> Thanks, Randy

If you have a folder with the code changes before the rebase please open a new PR with that, and see how that comes out.
